runtime/doc/ui.txt

```diff
@@ -576,6 +581,36 @@ because 'winhighlight' was used. UI items will be transmitted, even if the
 highlight group is cleared, so `ui_name` can always be used to reliably identify
 screen elements, even if no attributes have been applied.
 
+==============================================================================
+Details about UI element types				       *ui-elementtype*
+
+Activated by the `ext_elementtype` |ui-option|.
+Activates |ui-linegrid| implicitly.
+
+When this is set the `element_tags` parameter of |ui-event-hl_attr_define|
+contains semantic information about the type of UI element displayed.
+
+`element_tags` is a bitfield integer, with one more of the following tags set:
+    `StatusBar`    Status bars
+    `WinBar`       Winbars
```

Comment:
This doesn't explain at all why this new extension exists. All of the UI "ext" things literally exist to emit info about various UI widgets ("elements"). So this seems like an [inner platform](https://en.wikipedia.org/wiki/Inner-platform_effect), i.e. it's emitting UI info inside of the UI info system...

Reply:
The rest of the documentation does not really motivate each extension's existence either, so I didn't add that information to the documentation. But I can add a short description if needed.

The motivation is in the PR description and in the linked issue, but here's a longer one.

The immediate problem for Neovide is that UI elements like winbars, and borders are not detectable by the existing UI extensions, at least not in the general case, when the highlight group is something different than `FloatBorder` or `WinBar`. And they are part of the same grid as the buffer, so without being able to detect them it's not possible to know which parts of the grid should scroll.

I originally requested for them to be represented as separate grids, so we wouldn't need to care, but that was turned down by @bfredl here https://github.com/neovim/neovim/issues/23610#issuecomment-1867430580

But when I started to think more about it, I could think about plenty of more use cases. Changing the mouse shape when hovering elements is one of them, and that needs to be responsive, so a roundtrip to nvim is not an option. That's also one example, where only giving information about the borders as suggested, is not enough. You might want to change the cursor when hovering over URL links in the middle of the window for example.

Another use case is to render some of these elements in a more native style. For example, if we know that the type is s a border, then we could skip rendering the outer edges(background) of the glyphs. But to do that we need to know that it's a border, it could also be regular buffer text at the edge of the window. This is an alternative to having separate heavyweight extensions for everything, like `ext_tabline`, which also requires taking over the layout and a lot of other things.

BTW, I think UI elements is the correct term here https://en.wikipedia.org/wiki/List_of_graphical_user_interface_elements, for example a border is not of the widget subtype. But some of the elements are indeed widgets.